src/pip/_internal/download.py

```diff
@@ -199,13 +199,23 @@ def handle_401(self, resp, **kwargs):
 
         # Add our new username and password to the request
         req = HTTPBasicAuth(username or "", password or "")(resp.request)
+        req.register_hook("response", self.handle_401_again)
 
         # Send our new request
         new_resp = resp.connection.send(req, **kwargs)
         new_resp.history.append(resp)
 
         return new_resp
 
+    def handle_401_again(self, resp, **kwargs):
+        # warn user that they provided incorrect credentials
+        if resp.status_code != 401:
+            return resp
+
+        logger.warning('401 Error, Credentials not correct for %s',
+                       resp.request.url)
+        return resp
+
```

Comment:
If I'm reading the Requests documentation correctly, you only need to return a value if you're retruning something different. So

```python
if resp.status_code == 410:
    logger.warning(...)
```

should be all you need (and it looks cleaner).